designs/2021-suppression-support/README.md

```diff
@@ -0,0 +1,504 @@
+
+More specifically, when traversing `configuredRules` in `runRules`, ESLint will not [skip disabled rules](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L893) if the `--track-suppressions` option used. Then the suppression info `{kind: "external", justification: "Globally disabled"}` will be put into [`createReportTranslator`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/linter.js#L924) and [`createProblem`](https://github.com/eslint/eslint/blob/81c60f4a8725738f191580646562d1dca7eee933/lib/linter/report-translator.js#L357) to create reports with suppressions for disabled rules. If a non-existing rule is switched to `off`, ESLint would keep the current behavior and would not throw errors. The non-existing rules would not be recorded in suppression information as well.
+
+NOTE that rules disabled inside of configuration files could not be enabled by inline directive comments (e.g., `/* eslint-enable no-undef */`). But the disabling directive comments for disabled rules will still be recorded as an entry of `suppressions`, as described in [What we need](#What%20we%20need).
+
+### Reserve directive suppression
+
+In the step of apply directive suppression, ESLint will not remove the problems that should be disabled by directive comments if it gets the `--track-suppressions` option. These problems will be reserved and added `{kind: "directive", justification: "Justification message example."}`.
+
+NOTE that the kind `directive` should be converted to `inSource` for SARIF according to [kind property](https://docs.oasis-open.org/sarif/sarif/v2.1.0/os/sarif-v2.1.0-os.html#_Toc34317735).
+
+### Modify `CLIEngine` and related formatters
+
+As mentioned, we are not going to change the current behavior of ESLint. Without `--no-inline-config`, ESLint should not report errors/warnings in CMD whether with `--track-suppressions` option or not. In `CLIEngine`, when `suppressions` of a message is not null or empty (which means this message should be suppressed and not be reported), this message would be skipped or ignored. Additionally, the related formatters, such as stylish, should also be modified.
+
+## Documentation
+
+Add examples and explanations of the option `--track-suppressions` to _User guide – Command Line Interface_.
+
+## Drawbacks
+
+- Additional contexts must be transferred across modules, such as `justification` in `directives` and `suppressions` in `LintMessage`.
+- Related formatters, such as stylish, should be modified to keep the current behavior.
+
+## Backwards Compatibility Analysis
+
+Basically, current users would not feel any difference, since the proposed feature would only work with the new option. Without the option `--track-suppressions`, any current behaviors would not be changed.
+
+## Alternatives
+
+No.
```

Comment:
As an Alternatives: you can create an eslint rule to report the usage of `eslint-disable-*`. [eslint-plugin-eslint-comments](https://github.com/mysticatea/eslint-plugin-eslint-comments) sounds like a good place to have.

Reply:
We'd like to track external suppressions as well. Reporting the usage of `eslint-disable-*` seems not all we want exactly.

Reply:
as external suppressions, you can just read the eslintrc, to find which rules were disabled:

```js
config.rules.filter([rulename, ruleOptions] => ruleOptions === "off" || ruleOptions === 0);
```

Reply:
I want to clarify that we are not going to track what rules are suppressed (or switched to 'off'). What we want is tracking violations (or warnings/errors) in the code. For example:

```javascript
a = 1;
console.log('hello');
b = 2;
```

```json
rules: {
  "no-undef": "error"
}
```
We'd like to get the suppression info of line 1 (`a = 1`) and line 3 (`b = 2`) instead of the message that the rule 'no-undef' is switched to 'off'. No errors or warnings would be output, but these suppressions are expected to be recorded. As described in the RFC, ESLint would not skip rules disabled in configuration files if `--track-suppressions` is applied.

Reply:
@Yiwei-Ding when *what* switches it to off? another config, or only an inline override comment?

Reply:
For inline suppressions, we would keep the current logic basically. Violations (or problems in ESLint) disabled by inline directive comments are expected to be tracked but would not cause errors/warnings if '_--track-suppressions_'.

For external suppressions (where rules are switched to 'off' in configuration files), ESLint would not skip rules switched to 'off' if '_--track-suppressions_' is used. All rules would be ran and the generated problems belonging to disabled rules would be tagged 'Globally disabled'. In this RFC, when we talk about something is switched to 'off', we are talking about external suppressions (rules are switched to 'off' in configuration files).

Does it answer your questions?

Reply:
thanks, I see.

so, the alternative can be creating a new eslintrc based the original:

```js
const base = require('./eslintrc.js');  // this is your original config

base.rules = find_disabled_rules(base);  // find global-disabled rules in the config
base.rules['no-eslint-disable'] = "error"; // an eslint rule to report all the usage "eslint-disable-*"

modules.exports = base;
```

thus, you can run `eslint -c ` to get all suppressions. does it match your use case?

Reply:
What happens when a rule is switched off (from a shared config) in a config file because it crashes eslint, or because it' very slow?

Reply:
@aladdin-add I didn't get your idea..

Engineers prefer to suppress warnings or errors that would break CI rather than fix them. These suppressions are inevitable but we want to reduce them as much as possible to improve code quality. That is why we are going to track these suppressions.

So in this RFC, suppressions are expected to output from formatters (JSON, SARIF or any else) instead of causing warnings/errors to break pipelines. `"no-eslint-disable": "error"` would report inline directive comments as errors, right? That is not what we want.